src/dag_union.rs

```diff
@@ -0,0 +1,109 @@
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations
+// under the License.
+
+use crate::digraph::PyDiGraph;
+use petgraph::algo;
+use petgraph::graph::EdgeIndex;
+use pyo3::prelude::*;
+use pyo3::Python;
+use std::cmp::Ordering;
+use std::collections::BTreeMap;
+use std::collections::HashSet;
+
+/// [Graph] Return a new PyDiGraph by forming a union from`a` and `b` graphs.
+///
+/// The algorithm has three phases:
+///  - adds all nodes from `b` to `a`. operates in O(n), n being number of nodes in `b`.
+///  - merges nodes from `b` over `a` given that:
+///     - `merge_nodes` is `true`. operates in O(n^2), n being number of nodes in `b`.
+///     - respective node in`b` and `a` share the same weight
+///  - adds all edges from `b` to `a`.
+///     - `merge_edges` is `true`
+///     - respective edge in`b` and `a` share the same weight
+///
+/// with the same weight in graphs `a` and `b` and merged those nodes.
+///
+/// The nodes from graph `b` will replace nodes from `a`.
+///
+///  At this point, only `PyDiGraph` is supported.
+pub fn union(
+    py: Python,
+    a: &PyDiGraph,
+    b: &PyDiGraph,
+    merge_nodes: bool,
+    merge_edges: bool,
+) -> PyResult<PyDiGraph> {
+    let first = &a.graph;
+    let second = &b.graph;
+    let mut combined = PyDiGraph {
+        graph: first.clone(),
+        cycle_state: algo::DfsSpace::default(),
+        check_cycle: false,
+        node_removed: false,
+    };
+    let mut node_map = BTreeMap::new();
```

Comment:
How does the performance with BTreeMap compare to using a HashMap? In my experience the `hashbrown::HashMap` is faster for workloads where it's mostly inserts and gets/contains from the map (which I think this is).

Reply:
fixed